### PR TITLE
GSW-1838 refactor: tick_bitmap

### PR DESCRIPTION
**Pull request by r3v4s**
### tick_bitmap
- divide calculate function
- add test for each function

### CODE REVIEW

File: pool/tick_bitmap.gno [hunk 1]
```diff
@@ -96,18 +85,30 @@ func (pool *Pool) tickBitmapNextInitializedTickWithInOneWord(
 	}
 }
 
-// bitwiseNotForUint256BitmapIndexing performs a bitwise NOT operation on the uint256 value.
-// This function is designed for use in bitmap indexing
-func bitwiseNotForUint256BitmapIndexing(x *u256.Uint) *u256.Uint {
-	// REF: https://stackoverflow.com/a/77071037
+func generateFullMaskForBitPos(bitPos uint) *u256.Uint {
+	// 1 << bitPos
+	one := u256.NewUint(1)
```
Comment (notJoon) [inline pool/tick_bitmap.gno:90]: very niptick:

```suggestion
	one := u256.One()
```

**Review by notJoon — DISMISSED** [review]

File: pool/tick_bitmap.gno [hunk 2]
```diff
@@ -31,11 +30,9 @@ func (pool *Pool) tickBitmapFlipTick(
...
 
 	wordPos, bitPos := tickBitmapPosition(tick / tickSpacing)
+	pool.tickBitmaps[wordPos] = pool.tickBitmaps[wordPos].NilToZero()
 
-	// 1 << bitPos
 	mask := new(u256.Uint).Lsh(u256.NewUint(1), uint(bitPos))
-
-	pool.tickBitmaps[wordPos] = pool.tickBitmaps[wordPos].NilToZero()
 	pool.tickBitmaps[wordPos] = new(u256.Uint).Xor(pool.tickBitmaps[wordPos], mask)
 }
```
Comment (dongwon8247) [inline pool/tick_bitmap.gno:38]: Improve nested logic in `tickBitmapNextInitializedTickWithInOneWord`. Break them as much as possible and divide them into other func if necessary

File: pool/tick_bitmap.gno [hunk 3]
```diff
@@ -31,11 +30,9 @@ func (pool *Pool) tickBitmapFlipTick(
 	}
 
 	wordPos, bitPos := tickBitmapPosition(tick / tickSpacing)
+	pool.tickBitmaps[wordPos] = pool.tickBitmaps[wordPos].NilToZero()
```
Comment (dongwon8247) [inline pool/tick_bitmap.gno:33]: I see `tickBitmaps` being used multiple times. Abstract it into a helper func?

File: pool/tick_bitmap.gno [hunk 4]
```diff
@@ -96,18 +85,30 @@ func (pool *Pool) tickBitmapNextInitializedTickWithInOneWord(
 	}
 }
 
-// bitwiseNotForUint256BitmapIndexing performs a bitwise NOT operation on the uint256 value.
-// This function is designed for use in bitmap indexing
-func bitwiseNotForUint256BitmapIndexing(x *u256.Uint) *u256.Uint {
-	// REF: https://stackoverflow.com/a/77071037
+func generateFullMaskForBitPos(bitPos uint) *u256.Uint {
+	// 1 << bitPos
```
Comment (dongwon8247) [inline pool/tick_bitmap.gno:89]: You can unify this `(1 << bitPos) - 1` comment. It's distracting.

**Review by dongwon8247 — CHANGES_REQUESTED** [review]

File: pool/tick_bitmap.gno (hunk 4 above)
Comment (r3v4s) [reply pool/tick_bitmap.gno]: 964aebbe02b53e6e0c716f58d20823933c188001

File: pool/tick_bitmap.gno (hunk 2 above)
Comment (r3v4s) [reply pool/tick_bitmap.gno:37]: 22d47515d930c650b5e6f43b8351b39bea1eb613

File: pool/tick_bitmap.gno (hunk 3 above)
Comment (r3v4s) [reply pool/tick_bitmap.gno]: 22d47515d930c650b5e6f43b8351b39bea1eb613

File: pool/tick_bitmap.gno [hunk 5]
```diff
@@ -51,63 +47,91 @@ func (pool *Pool) tickBitmapNextInitializedTickWithInOneWord(
...
-		return next, initialized
+// getMaskBit gets a mask with all bits set to 1 up to the given bit position if lte is true,
+// or a mask with all bits set to 1 below the given bit position if lte is false
+func getMaskBit(bitPos uint, lte bool) *u256.Uint {
+	if lte {
+		return generateFullMaskForBitPos(bitPos)
+	}
+
+	return generateMaskClearingBitsBelow(bitPos)
+}
```
Comment (onlyhyde) [inline pool/tick_bitmap.gno:110]:
```suggestion
var bmap = map[bool]uint{
	true:  1,
	false: 0,
}

// getMaskBit gets a mask with all bits set to 1 up to the given bit position if lte is true,
// or a mask with all bits set to 1 below the given bit position if lte is false
func getMaskBit(bitPos uint, lte bool) *u256.Uint {
	shifted := new(u256.Uint).Lsh(u256.One(), bitPos+bmap[lte])
	mask := new(u256.Uint).Sub(shifted, u256.One())

	if !lte {
		return new(u256.Uint).Not(mask)
	}
	return mask
}
```

Split function names are understood differently than they are intended.
We modified the proposed code to compute in unison.

File: pool/tick_bitmap_test.gno [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+}
+
+func TestTickBitmapFlipTick(t *testing.T) {
+	tests := []struct {
+		name        string
+		tick        int32
+		tickSpacing int32
+		shouldPanic bool
+	}{
+		{
```
Comment (onlyhyde) [inline pool/tick_bitmap_test.gno:56]: Need to add negative value test case

**Review by onlyhyde — CHANGES_REQUESTED** [review]

File: pool/tick_bitmap.gno [hunk 7]
```diff
@@ -51,63 +47,91 @@ func (pool *Pool) tickBitmapNextInitializedTickWithInOneWord(
...
+	minusOne := new(u256.Uint).Sub(shifted, u256.One())
 
-		pool.tickBitmaps[wordPos] = pool.tickBitmaps[wordPos].NilToZero()
-		masked := new(u256.Uint).And(pool.tickBitmaps[wordPos], mask)
+	return new(u256.Uint).Add(minusOne, shifted)
+}
 
-		initialized := !(masked.IsZero())
+// generateMaskClearingBitsBelow generates a mask with all bits set to 1 below the given bit position
+func generateMaskClearingBitsBelow(bitPos uint) *u256.Uint {
```
Comment (notJoon) [inline pool/tick_bitmap.gno:69]: maybe you can reduce the number of object creations

```go
func generateMaskClearingBitsBelow(bitPos uint) *u256.Uint {
    result := new(u256.Uint)
    result.Lsh(u256.One(), bitPos)
    result.Sub(result, u256.One())
    return result.Not(result)
}
```

**Review by notJoon — APPROVED** [review]
LGTM

File: pool/tick_bitmap_test.gno (hunk 6 above)
Comment (r3v4s) [reply pool/tick_bitmap_test.gno:56]: c4b5a28711e09d74d36008fae5faf45a3acaa928

**Review by notJoon — APPROVED** [review]

**Comment by dongwon8247** [comment]
To be merged after @onlyhyde's review

**Review by dongwon8247 — APPROVED** [review]

**Review by onlyhyde — APPROVED** [review]

**Pull request merged**